turbopack/crates/turbo-tasks-backend/src/utils/sharded.rs

```diff
@@ -0,0 +1,51 @@
+use std::hash::{BuildHasher, BuildHasherDefault, Hash};
+
+use parking_lot::{Mutex, MutexGuard};
+use rustc_hash::FxHasher;
+
+pub struct Sharded<T, H = BuildHasherDefault<FxHasher>> {
+    data: Box<[Mutex<T>]>,
+    hasher: H,
+    bitmask: u16,
+}
```

Comment:
Yeah, I considered one of these lock free datastructures (I used some concurrent-queue before in the old backend), but it was missing an important feature that I needed: It is not able to atomically pop ALL items. Basically `take` the whole queue/list. So `Shared` is using the same technique as `DashMap` which seem to work really well and performant.

Reply:
> I used some concurrent-queue before in the old backend
> It is not able to atomically pop ALL items

It looks like `MARK_BIT` that `.close()`/`.is_closed` in `concurrent-queue` uses `SeqCst`? https://github.com/smol-rs/concurrent-queue/blob/e874f701f8413de01948c4903f894e5c845d8950/src/unbounded.rs#L391-L402

I think if you close the queue first, [`concurrent_queue::ConcurrentQueue`'s `try_iter` method](https://docs.rs/concurrent-queue/latest/concurrent_queue/struct.ConcurrentQueue.html#method.try_iter) _should_ atomically return all the elements? (Note: `TryIter::next()` is just calling `.pop()`)

If you want to atomically swap out the entire `ConcurrentQueue` for a new one when you do that, you could use a `RwLock` around the entire queue (preferably an [eventually fair](https://docs.rs/parking_lot/latest/parking_lot/type.RwLock.html#fairness) one?).

Reply:
You should just get this landed, and not block on this. However...

I was thinking about this a bit more, and have some ideas on how (if we want to keep `Sharded`) this might be improved.

Since we don't actually care what shard data ends up in (unlike `DashMap`, which does), we can:
- Pick a random shard, rather than hashing. This is probably a little faster.
- If the shard is locked (using `Mutex::try_lock`), just go to the next shard (some limited number of attempts), don't bother trying to wait to acquire the lock.
- Since lock contention shouldn't matter as much (we can just go to the next shard), decrease the total number of shards.